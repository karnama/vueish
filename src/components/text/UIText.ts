/**
 * Assert the key from the KeyboardEvent
 * was of an allowed type.
 */
export function onlyNumber(event: KeyboardEvent): void {
    const pattern = new RegExp('(?=\\d|\\.|(Backspace)|(ArrowLeft)|(ArrowRight)|(Delete))');

    if (!pattern.test(event.key)) {
        event.preventDefault();
    }
}
