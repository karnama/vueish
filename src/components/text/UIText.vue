<template>
    <div class="ui-text">
        <label :for="$attrs.id ?? $attrs.name" class="text-sm text-gray-500">
            {{ label }}
        </label>

        <div class="relative group">
            <div class="flex overflow-hidden">
                <span v-if="prefix ?? $slots.prefix" class="prefix mr-1">
                    <slot name="prefix">
                        {{ prefix }}
                    </slot>
                </span>
                <input :id="$attrs.id ?? $attrs.name"
                       v-bind="$attrs"
                       ref="input"
                       class="flex-1 appearance-none bg-transparent transition-border-color leading-tight
                       focus:outline-none rounded-none transition-text-color pb-2 disabled:cursor-not-allowed
                       disabled:text-gray-400"
                       :value="modelValue"
                       @input="$emit('update:modelValue', $event.target.value)"
                       @keydown="handleKeydown">

                <span v-if="suffix ?? $slots.suffix"
                      class="suffix mr-1 right-0"
                      :class="{'right-5': isNumber}">
                    <slot name="suffix">
                        {{ suffix }}
                    </slot>
                </span>

                <svg v-if="$attrs.disabled === true || $attrs.disabled === ''"
                     class="h-5 w-5 text-gray-400 right-0 top-1"
                     xmlns="http://www.w3.org/2000/svg"
                     viewBox="0 0 20 20"
                     fill="currentColor">
                    <path fill-rule="evenodd"
                          d="M5 9V7a5 5 0 0110 0v2a2 2 0 012 2v5a2 2 0 01-2
                             2H5a2 2 0 01-2-2v-5a2 2 0 012-2zm8-2v2H7V7a3 3 0 016 0z"
                          clip-rule="evenodd" />
                </svg>

                <svg v-else-if="!noClear && modelValue"
                     class="clear-icon h-5 w-5 cursor-pointer right-0 top-1 group-hover:opacity-100 transition-opacity
                            text-gray-500 relative -mr-5 group-hover:mr-0 -mt-1 transition-spacing"
                     xmlns="http://www.w3.org/2000/svg"
                     viewBox="0 0 20 20"
                     fill="currentColor"
                     @click="clearInput">
                    <path fill-rule="evenodd"
                          d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293
                             4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0
                             01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
                          clip-rule="evenodd" />
                </svg>
            </div>

            <div class="absolute w-full border-b left-0 bottom-0" />
        </div>
    </div>
</template>

<script lang="ts">
import { computed, defineComponent, ref } from 'vue';
import { SetupReturn } from '../../types';
import {
    autofocus,
    label,
    prefix,
    suffix,
    autofocusElement,
    useClearModelValue,
    noClear
} from '../../composables/input/input';
import { onlyNumber } from './UIText';

export default defineComponent({
    name: 'UIText',

    inheritAttrs: false,

    props: {
        modelValue: {
            type: [String, Number],
            required: true
        },

        prefix,
        suffix,
        label,
        autofocus,
        noClear
    },

    emits: ['update:modelValue'],

    setup(props, { emit, attrs }): SetupReturn {
        const input = ref<HTMLInputElement>();
        const clearInput = useClearModelValue(emit);
        const isNumber = computed(() => attrs.type === 'number');
        const handleKeydown = (event: KeyboardEvent) => isNumber.value && onlyNumber(event);
        autofocusElement(props.autofocus, input);

        return {
            input,
            clearInput,
            handleKeydown,
            isNumber
        };
    }
});
</script>

<style scoped>
/*https://stackoverflow.com/a/4298216*/
input:disabled::-webkit-outer-spin-button,
input:disabled::-webkit-inner-spin-button {
    /* display: none; <- Crashes Chrome on hover */
    -webkit-appearance: none;
    margin: 0; /* <-- Apparently some margin are still there even though it's hidden */
}

input:disabled[type=number] {
    -moz-appearance: textfield; /* Firefox */
}

.prefix, .suffix {
    line-height: 1.28;
}
</style>
