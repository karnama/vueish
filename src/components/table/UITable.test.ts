import type { Column, Row } from '@components/table/UITableTypes';
import { mount } from '@vue/test-utils';
import UITable from '@components/table/UITable.vue';
import { snakeCase } from 'lodash-es';
import UIText from '@components/text/UIText.vue';
import { nextTick, h } from 'vue';
import { orderBy } from 'lodash-es';

const headers: Readonly<Column[]> = [
    {
        header: 'Letter',
        rowProperty: 'letter',
        sortable: true,
        prefix: '(',
        suffix: ')'
    },
    {
        rowProperty: 'number',
        sortable: true,
        suffix: '%'
    }
];
const rows: Readonly<Row[]> = [
    {
        letter: 'b',
        number: 2,
        isSelectable: false
    },
    {
        letter: 'c',
        number: 3
    },
    {
        letter: 'a',
        number: 1
    }
];

const selectorMap = {
    search: '#search',
    rows: 'tbody > tr',
    headers: 'thead > tr.hidden.bg-gray-100 > th.py-6.text-left.px-4.uppercase',
    checkboxes: 'tbody > tr > td.px-2 input',
    topCheckbox: 'thead > tr.hidden.bg-gray-100 > th.py-6.px-2 > span.mx-auto input'
} as const;

function titleCase(str: string): string {
    return snakeCase(str) // Title Case
        .split('_')
        .reduce((previous: string, next: string) =>
            previous + ' ' + next.charAt(0).toUpperCase() + next.slice(1));
}

describe('UITable', () => {
    it('should display th given headers and rows', () => {
        const wrapper = mount(UITable, {
            props: {
                rows,
                headers
            }
        });

        const ths = wrapper.findAll(selectorMap.headers);
        expect(ths).toHaveLength(headers.length);
        headers.forEach(header => {
            const head = header.header ?? titleCase(header.rowProperty);
            expect(ths.map(th => th.text())).toContain(head);
        });

        const trs = wrapper.findAll(selectorMap.rows);
        expect(trs).toHaveLength(rows.length);
        rows.forEach((row, index) => {
            expect(trs[index].text()).toContain(row.number);
        });
        wrapper.unmount();
    });

    it('should figure out the header if not given', () => {
        const wrapper = mount(UITable, {
            props: {
                rows,
                headers
            }
        });

        expect(wrapper.html()).toContain(titleCase(headers[1].rowProperty));
        wrapper.unmount();
    });

    it('should display suffixes and prefixes when given', () => {
        const wrapper = mount(UITable, {
            props: {
                rows,
                headers
            }
        });

        const trs = wrapper.findAll(selectorMap.rows);
        expect(trs).toHaveLength(rows.length);
        rows.forEach((row, index) => {
            expect(trs[index].text()).toContain(String(row.number) + '%');
        });
        wrapper.unmount();
    });

    describe('search', () => {
        it('should only show the relevant rows on search', async () => {
            jest.useFakeTimers();
            const wrapper = mount(UITable, {
                props: {
                    rows,
                    headers,
                    search: true
                }
            });

            const search = wrapper.findComponent(UIText);
            await search.find('input').setValue('1');
            jest.runAllTimers(); // search is debounced
            await nextTick(); // wait for dom updates

            const searchResults = wrapper.findAll(selectorMap.rows);
            expect(searchResults).toHaveLength(1);
            expect(searchResults[0].text()).not.toContain('2');
            jest.useRealTimers();
            wrapper.unmount();
        });

        it('should search by the provided function if given', async () => {
            jest.useFakeTimers();
            const wrapper = mount(UITable, {
                props: {
                    rows,
                    headers,
                    search: (row: Row, term: string) => String(row.letter) + String(row.number) === term
                }
            });

            const search = wrapper.findComponent(UIText);
            await search.find('input').setValue('a1');
            jest.runAllTimers(); // search is debounced
            await nextTick(); // wait for dom updates

            const searchResults = wrapper.findAll(selectorMap.rows);
            expect(searchResults).toHaveLength(1);
            jest.useRealTimers();
            wrapper.unmount();
        });
    });

    describe('sorting', () => {
        it('should order the columns by ascending on click', async () => {
            const wrapper = mount(UITable, {
                props: {
                    rows,
                    headers
                }
            });

            const trs = wrapper.findAll(selectorMap.rows);

            // prove that they're in the order they were given
            rows.forEach((row, index) => {
                expect(trs[index].text()).toContain(row.letter);
            });

            const th = wrapper.findAll(selectorMap.headers).filter(dw => dw.text() === headers[0].header)[0]; // Letter
            await th.trigger('click');

            const orderedRows = orderBy(rows, ['letter']);
            wrapper.findAll(selectorMap.rows).forEach((tr, index) => {
                expect(tr.text()).toContain(orderedRows[index].letter);
            });
            wrapper.unmount();
        });

        it('should order the columns by descending on two clicks', async () => {
            const wrapper = mount(UITable, {
                props: {
                    rows,
                    headers
                }
            });

            // prove that they're in the order they were given
            wrapper.findAll(selectorMap.rows).forEach((tr, index) => {
                expect(tr.text()).toContain(rows[index].letter);
            });

            const th = wrapper.findAll(selectorMap.headers)
                .filter(dw => dw.text() === titleCase(headers[1].rowProperty))[0]; // Number
            await th.trigger('click');
            await th.trigger('click');

            const descOrderedRows = orderBy(rows, ['number'], ['desc']);
            wrapper.findAll(selectorMap.rows).forEach((tr, index) => {
                expect(tr.text()).toContain(descOrderedRows[index].letter);
            });
            wrapper.unmount();
        });

        it('should order the columns as it originally was on three clicks', async () => {
            const wrapper = mount(UITable, {
                props: {
                    rows,
                    headers
                }
            });

            // prove that they're in the order they were given
            wrapper.findAll(selectorMap.rows).forEach((tr, index) => {
                expect(tr.text()).toContain(rows[index].letter);
            });

            const th = wrapper.findAll(selectorMap.headers)
                .filter(dw => dw.text() === titleCase(headers[1].rowProperty))[0]; // Number
            await th.trigger('click');
            await th.trigger('click');
            await th.trigger('click');

            wrapper.findAll(selectorMap.rows).forEach((tr, index) => {
                expect(tr.text()).toContain(rows[index].letter);
            });
            wrapper.unmount();
        });

        it('should order the columns with the provided sort function', async () => {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const sortByFunc = jest.fn((row, _direction) => row.number);
            const wrapper = mount(UITable, {
                props: {
                    rows,
                    headers: [headers[0], {
                        ...headers[1],
                        sortByFunc
                    }]
                }
            });

            // prove that they're in the order they were given
            wrapper.findAll(selectorMap.rows).forEach((tr, index) => {
                expect(tr.text()).toContain(rows[index].letter);
            });

            const th = wrapper.findAll(selectorMap.headers)
                .filter(dw => dw.text() === titleCase(headers[1].rowProperty))[0]; // Number
            await th.trigger('click');

            let orderedRows = orderBy(rows, ['number']);
            expect(sortByFunc).toHaveBeenCalledTimes(rows.length);
            // partial match as not all attributes are required but in component are hydrated
            expect(sortByFunc).toHaveBeenLastCalledWith(expect.objectContaining(rows[rows.length - 1]));

            wrapper.findAll(selectorMap.rows).forEach((tr, index) => {
                expect(tr.text()).toContain(orderedRows[index].number);
            });

            await th.trigger('click');

            orderedRows = orderBy(rows, ['number'], ['desc']);
            expect(sortByFunc).toHaveBeenLastCalledWith(expect.objectContaining(rows[rows.length - 1]));
            wrapper.findAll(selectorMap.rows).forEach((tr, index) => {
                expect(tr.text()).toContain(orderedRows[index].letter);
            });
            wrapper.unmount();
        });

        it('should disable sorting when prop given', async () => {
            const wrapper = mount(UITable, {
                props: {
                    rows,
                    headers,
                    noSort: true
                }
            });

            // prove that they're in the order they were given
            wrapper.findAll(selectorMap.rows).forEach((tr, index) => {
                expect(tr.text()).toContain(rows[index].letter);
            });

            const th = wrapper.findAll(selectorMap.headers)
                .filter(dw => dw.text() === titleCase(headers[1].rowProperty))[0]; // Number
            await th.trigger('click');

            // it's still in the same order
            wrapper.findAll(selectorMap.rows).forEach((tr, index) => {
                expect(tr.text()).toContain(rows[index].letter);
            });
            wrapper.unmount();
        });
    });

    describe('selecting', () => {
        it('should v-model selection', async () => {
            const wrapper = mount(UITable, {
                props: {
                    rows,
                    headers,
                    modelValue: null,
                    showSelect: true
                }
            });

            const checkbox = wrapper.findAll(selectorMap.checkboxes)[0];
            await checkbox.trigger('click');
            await nextTick();

            // the first emit comes from the UICheckbox
            const payload = wrapper.emitted('update:modelValue')[1] as [Row];
            expect(payload[0].letter).toBe(rows[1].letter); // first is not selectable
            wrapper.unmount();
        });

        it('should v-model multiple selections', async () => {
            const wrapper = mount(UITable, {
                props: {
                    rows,
                    headers,
                    modelValue: null,
                    showSelect: true,
                    multiSelect: true
                }
            });

            const checkboxes = wrapper.findAll(selectorMap.checkboxes);
            await checkboxes[0].trigger('click');
            await nextTick();
            await checkboxes[1].trigger('click');
            await nextTick();

            // filter out the checkbox events
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
            const events = (wrapper.emitted('update:modelValue') as unknown[][])
                .filter(argumentArr => argumentArr.length === 1 && typeof argumentArr[0] !== 'boolean') as Row[][][];

            // console.log(events[events.length - 1]);
            // const sample = events[events.length - 1].map(row => row.isSelectable ? row.letter : null);
            // expect(sample).toStrictEqual(rows.map(row => row.isSelectable ? row.letter : null));
        });

        it.todo('should select all on clicking the top checkbox');

        it.todo('should not allow selecting a row if row is set to not be selectable');
    });

    describe('slots', () => {
        it('should accept footer, empty and header slots', async () => {
            const wrapper = mount(UITable, {
                props: {
                    rows,
                    headers
                },
                slots: {
                    footer: '<div id="footer-content" />',
                    empty: '<div id="empty-content" />',
                    header: (params) => h('div', { id: params.header.rowProperty }, JSON.stringify(params.header))
                }
            });

            expect(wrapper.find('#footer-content').exists()).toBe(true);
            headers.forEach(header => {
                // only look for the headers in the thead
                const headerWrapper = wrapper.find(selectorMap.headers + ' #' + header.rowProperty);
                expect(headerWrapper.exists()).toBe(true);

                // it was given the slotProps
                expect(JSON.parse(headerWrapper.text()).rowProperty).toBe(header.rowProperty);
            });

            await wrapper.setProps({ rows: [] });
            expect(wrapper.find('#empty-content').exists()).toBe(true);
            wrapper.unmount();
        });

        it('should dynamically accept slots based on header row property', () => {
            const slots = headers.reduce((
                previous: Record<string, string|((params: Record<string, any>) => ReturnType<typeof h>)>,
                next
            ) => {
                previous[next.rowProperty] = (params) => h(
                    'div',
                    { id: String(next.rowProperty) + '-content' },
                    JSON.stringify(params.row)
                );
                return previous;
            }, {});

            const wrapper = mount(UITable, {
                props: {
                    rows,
                    headers
                },
                slots
            });

            Object.keys(slots).forEach(slot => {
                const rowWrapper = wrapper.find(selectorMap.rows + ' #' + slot + '-content');

                expect(JSON.stringify(rows)).toContain(rowWrapper.text());
            });
            wrapper.unmount();
        });
    });
});
