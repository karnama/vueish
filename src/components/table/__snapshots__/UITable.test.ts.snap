// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UITable should display correctly 1`] = `
<section
  class="shadow dark:shadow-lg text-color bg-white dark:bg-gray-600 relative overflow-x-auto border border-gray-200 dark:border-gray-500 rounded"
>
  <table
    class="flex sm:table flex-col border-collapse w-full table-auto relative"
  >
    <thead
      class="sticky top-0 bg-gray-50 dark:bg-gray-650 border-b border-b-gray-300 dark:border-b-gray-500 shadow-sm dark:shadow-lg sm:!shadow-none z-10"
    >
      <!--v-if-->
      <tr
        class="hidden sm:table-row"
      >
        <!--v-if-->
        
        <th
          class="cursor-pointer hover:bg-gray-300 dark:hover:bg-gray-500 py-6 px-8 text-left uppercase font-light text-color-muted text-sm select-none group transition-colors"
        >
          <span
            class="flex items-center justify-between"
          >
            
            <span
              class="font-bold"
            >
              Letter
            </span>
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
        </th>
        <th
          class="cursor-pointer hover:bg-gray-300 dark:hover:bg-gray-500 py-6 px-8 text-left uppercase font-light text-color-muted text-sm select-none group transition-colors"
        >
          <span
            class="flex items-center justify-between"
          >
            
            <span
              class="font-bold"
            >
              number
            </span>
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
        </th>
        
        <!--v-if-->
      </tr>
    </thead>
    <tbody
      class="space-y-5 sm:space-y-0 divide-y"
    >
      
      <tr
        class="flex flex-col flex-no-wrap sm:table-row border-gray-200 dark:border-gray-500"
      >
        <!--v-if-->
        
        <td
          class="hover-cell-letter flex flex-row flex-nowrap items-center p-0 sm:table-cell"
          data-col="[object Object]"
          data-column="letter"
        >
          <span
            class="flex items-center justify-between sm:hidden content font-bold flex-none transition-colors select-none group cursor-pointer hover:bg-gray-300 p-4"
            role="rowheader"
            style="min-height: 40px; width: 140px;"
          >
            
            Letter
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
          <span
            class="block px-8 py-5"
          >
            
            (b)
            
          </span>
        </td>
        <td
          class="hover-cell-number flex flex-row flex-nowrap items-center p-0 sm:table-cell"
          data-col="[object Object]"
          data-column="number"
        >
          <span
            class="flex items-center justify-between sm:hidden content font-bold flex-none transition-colors select-none group cursor-pointer hover:bg-gray-300 p-4"
            role="rowheader"
            style="min-height: 40px; width: 140px;"
          >
            
            number
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
          <span
            class="block px-8 py-5"
          >
            
            2%
            
          </span>
        </td>
        
        <!--v-if-->
      </tr>
      <tr
        class="flex flex-col flex-no-wrap sm:table-row border-gray-200 dark:border-gray-500"
      >
        <!--v-if-->
        
        <td
          class="hover-cell-letter flex flex-row flex-nowrap items-center p-0 sm:table-cell"
          data-col="[object Object]"
          data-column="letter"
        >
          <span
            class="flex items-center justify-between sm:hidden content font-bold flex-none transition-colors select-none group cursor-pointer hover:bg-gray-300 p-4"
            role="rowheader"
            style="min-height: 40px; width: 140px;"
          >
            
            Letter
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
          <span
            class="block px-8 py-5"
          >
            
            (c)
            
          </span>
        </td>
        <td
          class="hover-cell-number flex flex-row flex-nowrap items-center p-0 sm:table-cell"
          data-col="[object Object]"
          data-column="number"
        >
          <span
            class="flex items-center justify-between sm:hidden content font-bold flex-none transition-colors select-none group cursor-pointer hover:bg-gray-300 p-4"
            role="rowheader"
            style="min-height: 40px; width: 140px;"
          >
            
            number
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
          <span
            class="block px-8 py-5"
          >
            
            3%
            
          </span>
        </td>
        
        <!--v-if-->
      </tr>
      <tr
        class="flex flex-col flex-no-wrap sm:table-row border-gray-200 dark:border-gray-500"
      >
        <!--v-if-->
        
        <td
          class="hover-cell-letter flex flex-row flex-nowrap items-center p-0 sm:table-cell"
          data-col="[object Object]"
          data-column="letter"
        >
          <span
            class="flex items-center justify-between sm:hidden content font-bold flex-none transition-colors select-none group cursor-pointer hover:bg-gray-300 p-4"
            role="rowheader"
            style="min-height: 40px; width: 140px;"
          >
            
            Letter
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
          <span
            class="block px-8 py-5"
          >
            
            (a)
            
          </span>
        </td>
        <td
          class="hover-cell-number flex flex-row flex-nowrap items-center p-0 sm:table-cell"
          data-col="[object Object]"
          data-column="number"
        >
          <span
            class="flex items-center justify-between sm:hidden content font-bold flex-none transition-colors select-none group cursor-pointer hover:bg-gray-300 p-4"
            role="rowheader"
            style="min-height: 40px; width: 140px;"
          >
            
            number
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
          <span
            class="block px-8 py-5"
          >
            
            1%
            
          </span>
        </td>
        
        <!--v-if-->
      </tr>
      
    </tbody>
    <!--v-if-->
  </table>
</section>
`;

exports[`UITable should display in its compact form 1`] = `
<section
  class="shadow dark:shadow-lg text-color bg-white dark:bg-gray-600 relative overflow-x-auto border border-gray-200 dark:border-gray-500 rounded"
  compact="true"
>
  <table
    class="flex sm:table flex-col border-collapse w-full table-auto relative"
  >
    <thead
      class="sticky top-0 bg-gray-50 dark:bg-gray-650 border-b border-b-gray-300 dark:border-b-gray-500 shadow-sm dark:shadow-lg sm:!shadow-none z-10"
    >
      <!--v-if-->
      <tr
        class="hidden sm:table-row"
      >
        <!--v-if-->
        
        <th
          class="cursor-pointer hover:bg-gray-300 dark:hover:bg-gray-500 py-6 px-8 text-left uppercase font-light text-color-muted text-sm select-none group transition-colors"
        >
          <span
            class="flex items-center justify-between"
          >
            
            <span
              class="font-bold"
            >
              Letter
            </span>
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
        </th>
        <th
          class="cursor-pointer hover:bg-gray-300 dark:hover:bg-gray-500 py-6 px-8 text-left uppercase font-light text-color-muted text-sm select-none group transition-colors"
        >
          <span
            class="flex items-center justify-between"
          >
            
            <span
              class="font-bold"
            >
              number
            </span>
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
        </th>
        
        <!--v-if-->
      </tr>
    </thead>
    <tbody
      class="space-y-5 sm:space-y-0 divide-y"
    >
      
      <tr
        class="flex flex-col flex-no-wrap sm:table-row border-gray-200 dark:border-gray-500"
      >
        <!--v-if-->
        
        <td
          class="hover-cell-letter flex flex-row flex-nowrap items-center p-0 sm:table-cell"
          data-col="[object Object]"
          data-column="letter"
        >
          <span
            class="flex items-center justify-between sm:hidden content font-bold flex-none transition-colors select-none group cursor-pointer hover:bg-gray-300 p-4"
            role="rowheader"
            style="min-height: 40px; width: 140px;"
          >
            
            Letter
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
          <span
            class="block px-8 py-5"
          >
            
            (b)
            
          </span>
        </td>
        <td
          class="hover-cell-number flex flex-row flex-nowrap items-center p-0 sm:table-cell"
          data-col="[object Object]"
          data-column="number"
        >
          <span
            class="flex items-center justify-between sm:hidden content font-bold flex-none transition-colors select-none group cursor-pointer hover:bg-gray-300 p-4"
            role="rowheader"
            style="min-height: 40px; width: 140px;"
          >
            
            number
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
          <span
            class="block px-8 py-5"
          >
            
            2%
            
          </span>
        </td>
        
        <!--v-if-->
      </tr>
      <tr
        class="flex flex-col flex-no-wrap sm:table-row border-gray-200 dark:border-gray-500"
      >
        <!--v-if-->
        
        <td
          class="hover-cell-letter flex flex-row flex-nowrap items-center p-0 sm:table-cell"
          data-col="[object Object]"
          data-column="letter"
        >
          <span
            class="flex items-center justify-between sm:hidden content font-bold flex-none transition-colors select-none group cursor-pointer hover:bg-gray-300 p-4"
            role="rowheader"
            style="min-height: 40px; width: 140px;"
          >
            
            Letter
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
          <span
            class="block px-8 py-5"
          >
            
            (c)
            
          </span>
        </td>
        <td
          class="hover-cell-number flex flex-row flex-nowrap items-center p-0 sm:table-cell"
          data-col="[object Object]"
          data-column="number"
        >
          <span
            class="flex items-center justify-between sm:hidden content font-bold flex-none transition-colors select-none group cursor-pointer hover:bg-gray-300 p-4"
            role="rowheader"
            style="min-height: 40px; width: 140px;"
          >
            
            number
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
          <span
            class="block px-8 py-5"
          >
            
            3%
            
          </span>
        </td>
        
        <!--v-if-->
      </tr>
      <tr
        class="flex flex-col flex-no-wrap sm:table-row border-gray-200 dark:border-gray-500"
      >
        <!--v-if-->
        
        <td
          class="hover-cell-letter flex flex-row flex-nowrap items-center p-0 sm:table-cell"
          data-col="[object Object]"
          data-column="letter"
        >
          <span
            class="flex items-center justify-between sm:hidden content font-bold flex-none transition-colors select-none group cursor-pointer hover:bg-gray-300 p-4"
            role="rowheader"
            style="min-height: 40px; width: 140px;"
          >
            
            Letter
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
          <span
            class="block px-8 py-5"
          >
            
            (a)
            
          </span>
        </td>
        <td
          class="hover-cell-number flex flex-row flex-nowrap items-center p-0 sm:table-cell"
          data-col="[object Object]"
          data-column="number"
        >
          <span
            class="flex items-center justify-between sm:hidden content font-bold flex-none transition-colors select-none group cursor-pointer hover:bg-gray-300 p-4"
            role="rowheader"
            style="min-height: 40px; width: 140px;"
          >
            
            number
            
            <i
              class="ml-2 transition opacity-0 group-hover:opacity-100"
            >
              <svg
                fill="currentColor"
                height="15px"
                viewBox="0 0 407.437 407.437"
                x="0px"
                xml:space="preserve"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <polygon
                  points="386.258,91.567 203.718,273.512 21.179,91.567 0,112.815 203.718,315.87 407.437,112.815 "
                />
              </svg>
            </i>
          </span>
          <span
            class="block px-8 py-5"
          >
            
            1%
            
          </span>
        </td>
        
        <!--v-if-->
      </tr>
      
    </tbody>
    <!--v-if-->
  </table>
</section>
`;
